x-pack/plugins/observability_onboarding/public/context/create_wizard_context.tsx

```diff
@@ -40,107 +46,136 @@ export function createWizardContext<
+
+  const routes = Object.entries(steps).reduce((acc, pair) => {
+    const [key, value] = pair as Entry<Record<StepKey, ComponentType>>;
+    return {
+      ...acc,
+      [stepRoute(key)]: {
+        exact: true,
+        handler: () =>
+          Page({
+            step: key,
+            Component: value,
+          }),
+      },
+    };
+  }, {});
+
+  function Page({
+    step,
+    Component,
+  }: {
+    step: StepKey;
+    Component: ComponentType;
+  }) {
+    const { setCurrentStep } = useWizard();
+    useEffect(() => {
+      setCurrentStep(step);
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [step]);
+
+    return <Component />;
+  }
+
   function Provider({
     children,
     onChangeStep,
     transitionDuration,
   }: {
-    children?: ReactNode;
+    children: ReactNode;
     onChangeStep?: (stepChangeEvent: {
       direction: 'back' | 'next';
       stepKey: StepKey;
       StepComponent: ComponentType;
     }) => void;
     transitionDuration?: number;
   }) {
-    const [step, setStep] = useState<StepKey>(initialStep);
-    const pathRef = useRef<StepKey[]>([initialStep]);
+    const history = useHistory();
+    const [step, setStep] = useState<StepKey>();
+    const pathRef = useRef<StepKey[]>([]);
     const usageRef = useRef<ReturnType<WizardContext<T, StepKey>['getUsage']>>({
       timeSinceStart: 0,
-      navEvents: [
-        { type: 'initial', step, timestamp: Date.now(), duration: 0 },
-      ],
+      navEvents: new Array<NavEvent<StepKey>>(),
     });
     const [state, setState] = useState<T>(initialState);
+
     return (
       <context.Provider
         value={{
-          CurrentStep: steps[step],
-          goToStep(stepKey: StepKey) {
+          setCurrentStep(stepKey: StepKey) {
             if (stepKey === step) {
               return;
             }
-            pathRef.current.push(stepKey);
-            const navEvents = usageRef.current.navEvents;
-            const currentNavEvent = navEvents[navEvents.length - 1];
-            const timestamp = Date.now();
-            currentNavEvent.duration = timestamp - currentNavEvent.timestamp;
-            usageRef.current.navEvents.push({
-              type: 'progress',
+
+            setStep(stepKey);
+            const stepVisited = pathRef.current.find((key) => key === stepKey);
+
+            pathRef.current = generatePath({
+              step: stepKey,
+              path: pathRef.current,
+            });
+
+            usageRef.current.navEvents = generateNavEvents({
+              type: stepVisited ? 'back' : 'progress',
               step: stepKey,
-              timestamp,
-              duration: 0,
+              navEvents: usageRef.current.navEvents as Array<NavEvent<StepKey>>,
             });
+
             if (onChangeStep) {
               onChangeStep({
-                direction: 'next',
+                direction: stepVisited ? 'back' : 'next',
```

Comment:
Should stepVisited be the criteria to consider it as going 'back' in steps?
What if there was a flow where we expect steps to be repeated, perhaps for additional configuration options:
![Screenshot 2023-07-13 at 23 23 16](https://github.com/elastic/kibana/assets/1967266/d25cec35-8041-4772-84a1-2c77aa876959)
I think we can keep it as is now, but we may have to revisit this for future designs.

Reply:
I get your point here, let's iterate over this in a future iteration. This would mean to pass the direction to the `setCurrentStep`